qp_klp/Step.py

```diff
@@ -0,0 +1,432 @@
+from collections import defaultdict
+from itertools import chain
+from json import dumps
+from metapool import KLSampleSheet
+from os import makedirs
+from os.path import join
+from sequence_processing_pipeline.ConvertJob import ConvertJob
+from sequence_processing_pipeline.FastQCJob import FastQCJob
+from sequence_processing_pipeline.GenPrepFileJob import GenPrepFileJob
+from sequence_processing_pipeline.PipelineError import PipelineError
+from sequence_processing_pipeline.Pipeline import Pipeline
+from sequence_processing_pipeline.QCJob import QCJob
+from subprocess import Popen, PIPE
+import pandas as pd
+
+
+class FailedSamplesRecord:
+    def __init__(self, output_dir, samples):
+        # because we want to write out the list of samples that failed after
+        # each Job is run, and we want to organize that output by project, we
+        # need to keep a running state of failed samples, and reuse the method
+        # to reorganize the running-results and write them out to disk.
+
+        self.output_path = join(output_dir, 'failed_samples.html')
+
+        # create an initial dictionary with sample-ids as keys and their
+        # associated project-name and status as values. Afterwards, we'll
+        # filter out the sample-ids w/no status (meaning they were
+        # successfully processed) before writing the failed entries out to
+        # file.
+        self.failed = {x.Sample_ID: [x.Sample_Project, None] for x in samples}
```

Comment:
is this really failed? At this stage it looks like all samples; it might be confusing for other devs.

Reply:
One option is to have: self.samples (all samples) and then another one with self.samples_used and self.failed can be the difference between both.

Reply:
I changed the name from self.failed to self.sample_state - you're right, we're initializing the member with the entire list of samples but we don't know that any of them are failed yet.